### DISCUSSION TITLE
appending a canvas to a SceneryStack node?

### DISCUSSION THREAD

**Discussion by dave-english** [scenerystack/community#54] [Q&A] [answered]
I'd like to embed an electrical arc effect in my Sim, built with SceneryStack.  I've studied the code from ([here](https://diwsi.github.io/jslightning/index.html),) and have some idea of how to insert it into my own code, except that the main function calls for a canvas.context object as one of is parameters. which I might get as:
```
var c = document.getElementById("myCanvas");
var ctx = c.getContext("2d");
```
But that assumes that "myCanvas" is already a hook in the HTML code.
Is there a particular type of SceneryStack node that has the properties of a Canvas, or a way to pass some property of the node to a function that expects a canvas.context?

Or should I be discouraged from taking this approach because it won't work in non-Canvas environments?

**Comment by pixelzoom** [comment, marked as answer]
See Scenery’s `CanvasNode`, and its usages in many PhET simulations, for example https://github.com/phetsims/acid-base-solutions/blob/main/js/common/view/ParticlesCanvasNode.ts.

Re non-Canvas environments… Support for Canvas is ubiquitous across browsers these days.  But if you have a specific (or unusual) platform in mind, you might check one of the many browser-compatibility tables available online.

Re code you code snippet with `document.getElementById`… One of Scenery’s primary roles is to provide a higher-level API that encapsulates the Document, freeing the developer from having to deal with things like feature compatibility across browsers.  As soon as you break that encapsulation by accessing the Document directly, you lose some of the benefits of scenery. So (at least with PhET simulation development) accessing the Document directly is generally discouraged and unnecessary — I’ve never had a need to do so.  If you find a feature that Scenery does not support, it would be preferable to make a feature request, rather go around Scenery.

**Reply by dave-english** [reply]
Thank you. This is helpful information. I'll take a look at the CanvasNode in the future. Meantime, I've decided to implement my electric arc graphic using Scenery's own classes, notably Shape, Path and Vector2, to avoid direct document access.

**Reply by pixelzoom** [reply]
In case there is any confusion... CanvasNode is also one of Scenery's own classes, and would avoid direct Document access.
